x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialRelatesFunction.java

```diff
@@ -0,0 +1,499 @@
+    }
+
+    protected static Component2D asLuceneComponent2D(DataType spatialDataType, Expression expression) {
+        Object result = expression.fold();
+        if (result instanceof String string) {
+            return asLuceneComponent2D(spatialDataType, SpatialCoordinateTypes.UNSPECIFIED.wktToGeometry(string));
+        } else if (result instanceof BytesRef bytesRef) {
+            return asLuceneComponent2D(spatialDataType, SpatialCoordinateTypes.UNSPECIFIED.wktToGeometry(bytesRef.utf8ToString()));
+        } else {
+            throw new IllegalArgumentException("Invalid spatial constant string: " + result);
+        }
+    }
+
+    protected static Component2D asLuceneComponent2D(DataType spatialDataType, Geometry geometry) {
+        if (EsqlDataTypes.isSpatialGeo(spatialDataType)) {
+            var luceneGeometries = LuceneGeometriesUtils.toLatLonGeometry(geometry, false, t -> {});
+            return LatLonGeometry.create(luceneGeometries);
+        } else {
+            var luceneGeometries = LuceneGeometriesUtils.toXYGeometry(geometry, t -> {});
+            return XYGeometry.create(luceneGeometries);
+        }
+    }
+
+    protected static GeometryDocValueReader asGeometryDocValueReader(DataType spatialDataType, Expression expression) throws IOException {
+        Object result = expression.fold();
+        if (result instanceof String string) {
+            return asGeometryDocValueReader(spatialDataType, SpatialCoordinateTypes.UNSPECIFIED.wktToGeometry(string));
+        } else if (result instanceof BytesRef bytesRef) {
+            return asGeometryDocValueReader(spatialDataType, SpatialCoordinateTypes.UNSPECIFIED.wktToGeometry(bytesRef.utf8ToString()));
+        } else {
+            throw new IllegalArgumentException("Invalid spatial constant string: " + result);
+        }
+    }
+
+    protected static GeometryDocValueReader asGeometryDocValueReader(DataType spatialDataType, Geometry geometry) throws IOException {
+        if (EsqlDataTypes.isSpatialGeo(spatialDataType)) {
+            return asGeometryDocValueReader(
+                new GeoShapeCoordinateEncoder(),
+                new GeoShapeIndexer(Orientation.CCW, "SpatialRelatesFunction"),
+                geometry
+            );
+        } else {
+            return asGeometryDocValueReader(
+                new CartesianShapeCoordinateEncoder(),
+                new CartesianShapeIndexer("SpatialRelatesFunction"),
+                geometry
+            );
+        }
+
+    }
+
+    protected static GeometryDocValueReader asGeometryDocValueReader(
+        CoordinateEncoder encoder,
+        ShapeIndexer shapeIndexer,
+        Geometry geometry
+    ) throws IOException {
+        GeometryDocValueReader reader = new GeometryDocValueReader();
+        CentroidCalculator centroidCalculator = new CentroidCalculator();
+        if (geometry instanceof Circle circle) {
+            // TODO: How should we deal with circles?
```

Comment:
You can use the CircleUtils in the geo lib to polygonize it but I think we should throw an error in this case as we do not support indexing circles anyway. shapeIndexer.indexShape(geometry) will throw an error,

Reply:
So, since both the centroid calculator and the shape indexer do not support circles, I now throw an exception directly here. Supporting shapes using CircleUtils could be considered as a future enhancement, either by itself, or as part of more advanced Curve support similar to PostGIS.